tools/upload_text.php

```diff
@@ -249,164 +217,152 @@
     // make reasonably sure script does not timeout on large file uploads
     set_time_limit(14400);
 
+    // Disable gzip compression so we can flush the buffer after each step
+    // in the process to give the user some progress details. Note that this
+    // doesn't necessarily work for all browsers.
+    apache_setenv('no-gzip', '1');
+
+    slim_header($title);
+    echo "<h1>$title</h1>";
+    echo "<h2>", sprintf("Project: %s", $project->nameofwork), "</h2>";
+
     // if files have been uploaded, process them and mangle the postcomments
-    $uploaded_file = validate_uploaded_file();
+    $returning_to_pool = ('return_1' == $stage || 'return_2' == $stage);
+    try
+    {
+        $filename = process_file($project, $indicator, $stage, $returning_to_pool);
+        // we've put the file in the right place.
+        // now let's deal with the postcomments.
+        // we construct the bit that's going to be added on to the existing postcomments.
+        // if we're returning to available, and the user hasn't loaded a file, and not
+        // entered any comments, we don't bother.
+        // Otherwise, we add a divider, time stamp, user name, and the name of the file
+        // "uploaded by" & "returned by" not translated since they go into postcomments rather than being viewed by the present user
+        $divider = "\n----------\n".date("Y-m-d H:i");
+        if ($filename) {
+            $divider .= " $filename uploaded by ";
+        }
+        else if ($returning_to_pool) {
+            $divider .= " returned by ";
+        }
+        else {
+            $divider .= " "; // this shouldn't actually happen
+        }
+        $divider .= $pguser."\n";
+        if (strlen($postcomments)>0 || $filename) {
+            $postcomments = $divider . $postcomments;
+        }
+        // note that $postcomments is used as a global variable inside do_state_change() inside project_transition()
 
-    $have_file = FALSE;
-    if (is_file($uploaded_file))
+        $error_msg = project_transition( $projectid, $new_state, $pguser, $extras );
+        if ($error_msg)
+        {
+            throw new FileUploadException($error_msg);
+        }
+
+        // special handling for smooth reading, which does not involve a state change
+        // so project_transition() will do nothing
+        // but still needs some changes recorded in project table
```

Comment:
Should `project table` be `project_events` table or `projects` table, or both? `smoothread_deadline` is stored in the `projects` table, but the assorted events recorded in the project history are recorded in `project_events`.

Reply:
> However, for example, if I attempt to upload a project for PPV, and neglect to choose a file (or choose a bad file), I'm sent back to the PP pool page, not to the project I was trying to upload for verification.
For upload_text.php I've changed the links to 'Go Back' when the operation has not succeeded.

Reply:
> Was there also an issue with there being no return link at all if an upload failed for non-resumables?

There was. Although it is not possible to do anything better after it has happened I've added a check for file size before the upload starts which should prevent it ever happening.

Reply:
> Should project table be project_events table or projects table, or both? smoothread_deadline is stored in the projects table, but the assorted events displayed in the project history are recorded in project_events.

Yes, it should be both. I've changed the comment.